File: java/src/main/java/com/adsbynimbus/openrtb/impression/Creative.java
package com.adsbynimbus.openrtb.impression;

/**
 * Defines constants shared between children of the {@link Impression} object
 */
public abstract class Creative {

    public static final String BID_FLOOR = "bidfloor"; //float
    public static final String POSITION = "pos"; // int
    public static final String MIME_TYPES = "mimes"; // String[]
    public static final String API = "api"; // int[]

    /* Supported APIS [See OpenRTB 2.5 Section 5.6] */
    public static final int VPAID_1 = 1;
    public static final int VPAID_2 = 2;
    public static final int MRAID_1 = 3;
    public static final int ORMMA = 4;
    public static final int MRAID_2 = 5;
    public static final int MRAID_3 = 6;

    //Ad Positions
    public static final int POSITION_UNKNOWN = 0;
    public static final int ABOVE_THE_FOLD = 1;
    //public static final int DEPRECATED = 2;
    public static final int BELOW_THE_FOLD = 3;
    public static final int HEADER = 4;
    public static final int FOOTER = 5;
    public static final int SIDEBAR = 6;
    public static final int FULL_SCREEN = 7;

    //Share API version
    public static final int OMID = 7;

    public static final int AUDIO_AD_AUTO_PLAY = 1;
    public static final int AUDIO_AD_USER_INITIATED = 2;
    public static final int EXPANDABLE_AUTOMATIC = 3;
    public static final int EXPANDABLE_USER_CLICK = 4;
    public static final int EXPANDABLE_USER_ROLLOVER = 5;
    public static final int BANNER_VIDEO_AUTO_PLAY = 6;
    public static final int BANNER_VIDEO_USER_INITIATED = 7;
    public static final int HAS_POPUP = 8;
    public static final int PROVOCATIVE_OR_SUGGESTIVE = 9;
    public static final int EXTREME_ANIMATION = 10;
    public static final int SURVEYS = 11;
    public static final int TEXT_ONLY = 12;
    public static final int USER_INTERACTIVE_AND_GAMES = 13;
    public static final int DIALOG_OR_ALERT_STYLE = 14;
    public static final int HAS_VOLUME_TOGGLE = 15;
    public static final int HAS_SKIP_BUTTON = 16;
    public static final int ADOBE_FLASH = 17;
}

File: kotlin/src/commonMain/kotlin/request/Creative.kt
package com.adsbynimbus.openrtb.request

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * API frameworks supported by the publisher for a creative
 *
 * [OpenRTB Section 5.6](https://www.iab.com/wp-content/uploads/2016/03/OpenRTB-API-Specification-Version-2-5-FINAL.pdf#page=51)
 */
public object Api {
    public const val Vpaid1: Int = 1
    public const val Vpaid2: Int = 2
    public const val Mraid1: Int = 3
    public const val Ormma: Int = 4
    public const val Mraid2: Int = 5
    public const val Mraid3: Int = 6
    public const val Omid: Int = 7
}

/**
 * Standard list of creative attributes that can describe an ad being served or serve as
 * restrictions of thereof.
 *
 * [OpenRTB Section 5.3](https://www.iab.com/wp-content/uploads/2016/03/OpenRTB-API-Specification-Version-2-5-FINAL.pdf#page=50)
 */
public object Attribute {
    public const val AudioAdAutoPlay: Int = 1
    public const val AudioAdUserInitiated: Int = 2
    public const val ExpandableAutomatic: Int = 3
    public const val ExpandableUserClick: Int = 4
    public const val ExpandableUserRollover: Int = 5
    public const val BannerVideoAutoPlay: Int = 6
    public const val BannerVideoUserInitiated: Int = 7
    public const val HasPopup: Int = 8
    public const val ProvocativeOrSuggestive: Int = 9
    public const val ExtremeAnimation: Int = 10
    public const val Surveys: Int = 11
    public const val TextOnly: Int = 12
    public const val UserInteractiveAndGames: Int = 13
    public const val DialogOrAlertStyle: Int = 14
    public const val HasVolumeToggle: Int = 15
    public const val HasSkipButton: Int = 16
    public const val AdobeFlash: Int = 17
}

/**
 * Describes the position of the ad as a relative measure of visibility or prominence. This OpenRTB
 * table has values derived from the Inventory Quality Guidelines (IQG). Values 4 - 7 apply to apps.
 *
 * [OpenRTB Section 5.4](https://www.iab.com/wp-content/uploads/2016/03/OpenRTB-API-Specification-Version-2-5-FINAL.pdf#page=51)
 */
public object Position {
    public const val POSITION_UNKNOWN: Int = 0
    public const val ABOVE_THE_FOLD: Int = 1
    public const val BELOW_THE_FOLD: Int = 3
    public const val HEADER: Int = 4
    public const val FOOTER: Int = 5
    public const val SIDEBAR: Int = 6
    public const val FULL_SCREEN: Int = 7
}
